kpi/tests/api/v2/test_api_assets.py

```diff
@@ -742,19 +749,55 @@ def verify_asset_file(self, response):
             response_dict['user__username'],
             self.current_username,
         )
+
+        # Some metadata properties are added when file is created.
+        # Let's compare without them
+        responsed_metadata = dict(response_dict['metadata'])
+
+        if not form_media:
+            # `filename` is only mandatory with form media files
+            responsed_metadata.pop('filename', None)
+
+        responsed_metadata.pop('mimetype', None)
+        responsed_metadata.pop('hash', None)
+
         self.assertEqual(
-            json.dumps(response_dict['metadata']),
+            json.dumps(responsed_metadata),
             posted_payload['metadata']
         )
-        for field in 'file_type', 'name':
+        for field in 'file_type', 'description':
             self.assertEqual(response_dict[field], posted_payload[field])
+
         # Content via the direct URL to the file
-        posted_payload['content'].seek(0)
-        expected_content = posted_payload['content'].read()
-        self.assertEqual(
-            self.get_asset_file_content(response_dict['content']),
-            expected_content
-        )
+        try:
+            posted_payload['content'].seek(0)
+        except KeyError:
+            pass
+        else:
+            expected_content = posted_payload['content'].read().encode()
+            self.assertEqual(
+                self.get_asset_file_content(response_dict['content']),
+                expected_content
+            )
+
+        try:
+            base64_encoded = posted_payload['base64Encoded']
+        except KeyError:
+            pass
```

Comment:
Actually, the serializer already validates this. There was already a test for 2 different upload methods at the same time.
I added another with no methods.

Reply:
Looks good. I see that a `KeyError` being raised when looking for `redirect_url` is the final backstop, so that `verify_asset_file()` won't silently return if neither binary, base64, nor a redirect URL has been provided.